src/main/java/com/minecolonies/coremod/entity/ai/mobs/EntityAIAttackRangedBowBarbarian.java

```diff
@@ -0,0 +1,196 @@
+
+import net.minecraft.entity.EntityLivingBase;
+import net.minecraft.entity.ai.EntityAIBase;
+import net.minecraft.init.Items;
+import net.minecraft.item.ItemBow;
+import net.minecraft.util.EnumHand;
+
+/**
+ * Class implementing the AI for the Archer Barbarians ranged attack.
+ */
+public class EntityAIAttackRangedBowBarbarian extends EntityAIBase
+{
+    private final AbstractArcherBarbarian entity;
+    private final double                  moveSpeedAmp;
+    private       int                     attackCooldown;
+    private final double                  maxAttackDistance;
+    private int attackTime = -1;
+    private int     seeTime;
+    private boolean strafingClockwise;
+    private boolean strafingBackwards;
+    private              int    strafingTime         = -1;
+    private final        int    ONE_MINUTE           = 60;
+    private final        int    TWENTY_SECONDS       = 20;
+    private final        double ONE_QUARTER_BLOCK    = 0.25D;
+    private final        float  HALF_BLOCK           = 0.5F;
+    private final        double THREE_QUARTERS_BLOCK = 0.75D;
+    private final        float  MAX_DEGREES          = 30.0F;
+    private static final int    MUTEX_BITS           = 3;
+
+    public EntityAIAttackRangedBowBarbarian(final AbstractArcherBarbarian archer, final double speedAmplifier, final int delay, final double maxDistance)
+    {
+        super();
+        this.entity = archer;
+        this.moveSpeedAmp = speedAmplifier;
+        this.attackCooldown = delay;
+        this.maxAttackDistance = maxDistance * maxDistance;
+        this.setMutexBits(MUTEX_BITS);
+    }
+
+    public void setAttackCooldown(final int coolDown)
+    {
+        this.attackCooldown = coolDown;
+    }
+
+    /**
+     * Returns whether the EntityAIBase should begin execution.
+     */
+    public boolean shouldExecute()
+    {
+        if (this.entity.getAttackTarget() == null)
+        {
+            return false;
+        }
+        return this.isBowInMainhand();
+    }
+
+    protected boolean isBowInMainhand()
+    {
+        return !this.entity.getHeldItemMainhand().isEmpty() && this.entity.getHeldItemMainhand().getItem() == Items.BOW;
+    }
+
+    /**
+     * Returns whether an in-progress EntityAIBase should continue executing
+     */
+    public boolean continueExecuting()
+    {
+        return (this.shouldExecute() || !this.entity.getNavigator().noPath()) && this.isBowInMainhand();
+    }
+
+    /**
+     * Execute a one shot task or start executing a continuous task
+     */
+    public void startExecuting()
+    {
+        super.startExecuting();
+        this.entity.setSwingingArms(true);
+    }
+
+    /**
+     * Resets the task
+     */
+    public void resetTask()
+    {
+        super.resetTask();
+        this.entity.setSwingingArms(false);
+        this.seeTime = 0;
+        this.attackTime = -1;
+        this.entity.resetActiveHand();
+    }
+
+    /**
+     * Updates the task
+     */
+    public void updateTask()
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') The Cyclomatic Complexity of this method "updateTask" is 21 which is greater than 15 authorized. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor this method to reduce its Cognitive Complexity from 44 to the 20 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)